sqlite.go

```diff
@@ -0,0 +1,530 @@
+}
+func (s *stmt) Exec(args []driver.Value) (driver.Result, error) { panic("unused") }
+func (s *stmt) Query(args []driver.Value) (driver.Rows, error)  { panic("unused") }
+
+func (s *stmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	if err := s.stmt.Reset(); err != nil {
+		return nil, s.reserr("Stmt.Query(Reset)", err)
+	}
+	defer s.stmt.ClearBindings()
+	defer s.stmt.Reset()
+	if err := s.bindAll(args); err != nil {
+		return nil, err
+	}
+	row, lastInsertRowID, changes, err := s.stmt.StepResult()
+	_ = row // TODO: return error if exec on query which returns rows?
+	if err != nil {
+		return nil, err
+	}
+	return stmtResult{lastInsertID: lastInsertRowID, rowsAffected: changes}, nil
+}
+
+type stmtResult struct {
+	lastInsertID int64
+	rowsAffected int64
+}
+
+func (res stmtResult) LastInsertId() (int64, error) { return res.lastInsertID, nil }
+func (res stmtResult) RowsAffected() (int64, error) { return res.rowsAffected, nil }
+
+func (s *stmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	if err := s.stmt.Reset(); err != nil {
+		return nil, s.reserr("Stmt.Query(Reset)", err)
+	}
+	if err := s.bindAll(args); err != nil {
+		return nil, err
+	}
+	return &rows{stmt: s}, nil
+}
+
+func (s *stmt) bindAll(args []driver.NamedValue) error {
+	for _, arg := range args {
+		if err := s.bind(arg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *stmt) bind(arg driver.NamedValue) error {
+	var debugName interface{} = arg.Ordinal
+	if arg.Name != "" {
+		debugName = arg.Name
+		index := s.stmt.BindParameterIndexSearch(arg.Name)
+		if index == 0 {
+			return &Error{
+				Code:  sqliteh.SQLITE_MISUSE,
+				Loc:   "Bind",
+				Query: s.query,
+				Msg:   fmt.Sprintf("unknown parameter name %q", arg.Name),
+			}
+		}
+		arg.Ordinal = index
+	}
+
+	// Start with obvious types, including time.Time before TextMarshaler.
+	found, err := s.bindBasic(debugName, arg.Ordinal, arg.Value)
+	if err != nil {
+		return err
+	} else if found {
+		return nil
+	}
+
+	if m, _ := arg.Value.(encoding.TextMarshaler); m != nil {
+		b, err := m.MarshalText()
+		if err != nil {
+			// TODO: modify Error to carry an error so we can %w?
+			return &Error{
+				Code:  sqliteh.SQLITE_MISUSE,
+				Loc:   "Bind",
+				Query: s.query,
+				Msg:   fmt.Sprintf("Bind:%v: cannot marshal %T: %v", debugName, arg.Value, err),
+			}
+		}
+		_, err = s.bindBasic(debugName, arg.Ordinal, b)
+		return err
+	}
+
+	// Look for named basic types or other convertible types.
+	val := reflect.ValueOf(arg.Value)
+	typ := reflect.TypeOf(arg.Value)
+	switch typ.Kind() {
+	case reflect.Bool:
+		b := int64(0)
+		if val.Bool() {
+			b = 1
+		}
+		_, err := s.bindBasic(debugName, arg.Ordinal, b)
+		return err
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		_, err := s.bindBasic(debugName, arg.Ordinal, val.Int())
+		return err
+	case reflect.Uint, reflect.Uint64:
```

Comment:
They can? This code is handling when someone passes in a uint64 directly. If they pass in a uint64 type which is a TextMarshaler, it's caught just above and handled. They can also make their own string and pass it in.

Not really sure what you want me to do here.

Reply:
Oh, I see. I missed the order of operations here, because I was flipping back and forth to bindBasic a lot. I still am tempted to try to make uint and uint64 just work, at least sometimes, but I guess that's a lost cause.